src/blob.rs

```diff
@@ -1,86 +1,258 @@
-    pub fn write(&mut self, _record: Record<K>) -> Result<(), ()> {
-        // @TODO implement
-        Ok(())
+#[derive(Debug)]
+pub struct ReadFuture {
+    file: fs::File,
+    key: Vec<u8>,
+    loc: Location,
+}
+
+impl Future for ReadFuture {
+    type Output = Result<Record>;
+
+    fn poll(self: Pin<&mut Self>, _waker: &Waker) -> Poll<Self::Output> {
+        let mut buf = vec![0u8; self.loc.size as usize];
+        self.file.read_at(&mut buf, self.loc.offset).unwrap();
+        Poll::Ready(Record::from_raw(&buf).map_err(Error::RecordError))
     }
+}
 
+impl Blob {
     /// # Description
-    /// Reads record data, yields `Ok(Vec<u8>)` if read successful,
-    /// otherwise - `Err`
-    // @TODO more useful result
-    pub fn read(&self, _key: &K) -> Result<Record<K>, ()> {
-        // @TODO implement
-        Ok(Record::new())
+    /// Creates new blob file
+    pub fn open_new(name: FileName) -> Result<Self> {
+        Ok(Self {
+            header: Header::new(),
+            file: fs::OpenOptions::new()
+                .create_new(true)
+                .append(true)
+                .read(true)
+                .open(name.as_path())
+                .map_err(Error::OpenNew)?,
+            name,
+            index: BTreeMap::new(),
+            current_offset: 0,
+        })
+    }
+
+    pub fn boxed(self) -> Box<Self> {
+        Box::new(self)
+    }
+
+    pub fn from_file(path: PathBuf) -> Result<Self> {
+        let file = fs::OpenOptions::new()
+            .create(false)
+            .append(true)
+            .read(true)
+            .open(&path)
+            .map_err(Error::FromFile)?;
+        let name = FileName::from_path(&path)?;
+        let len = file.metadata().map_err(Error::FromFile)?.len();
+        // @TODO Scan existing file to create index
+        Ok(Self {
+            header: Header::new(),
+            file,
+            name,
+            index: Default::default(),
+            current_offset: len,
+        })
+    }
+
+    pub fn write(&mut self, record: &mut Record) -> Result<WriteFuture> {
+        record.set_blob_offset(self.current_offset);
+        self.index
+            .insert(record.key().to_vec(), record.header().clone());
+        let buf = record.to_raw();
+        self.current_offset += buf.len() as u64;
+        Ok(WriteFuture {
+            file: self.file.try_clone().map_err(Error::CloneFd)?,
+            buf: Some(buf),
+            current_offset: self.current_offset,
+        })
+    }
+
+    pub fn read<K>(&self, key: &K) -> Result<ReadFuture>
+    where
+        K: AsRef<[u8]> + Ord,
+    {
+        Ok(ReadFuture {
+            file: self.file.try_clone().map_err(Error::CloneFd)?,
+            key: key.as_ref().to_vec(),
+            loc: self.locate(&key)?,
+        })
+    }
+
+    fn locate<K>(&self, key: &K) -> Result<Location>
+    where
+        K: AsRef<[u8]> + Ord,
+    {
+        let header = self.index.get(key.as_ref()).ok_or(Error::NotFound)?;
```

Comment:
You don't read `index` when you open an existing file. This will result in `Error::NotFound`, even when the record is inside the BLOB.

Reply:
rename locate - lookup

Reply:
Read index from existing file is the next issue
Rename done